File: surface/widgets/compose/components/widgets/form-controls/Stepper.kt
// ============================================================
// Clef Surface Compose Widget — Stepper
//
// Compact increment/decrement control. Renders as a Row with
// [-] value [+] icon buttons flanking the current value.
// Respects min, max, and step constraints. Maps the
// stepper.widget anatomy (root, label, decrementButton,
// value, incrementButton) to Material 3 IconButton + Text.
// ============================================================

package clef.surface.compose.components.widgets.formcontrols

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

// --------------- Component ---------------

@Composable
fun Stepper(
    value: Int,
    onValueChange: (Int) -> Unit,
    modifier: Modifier = Modifier,
    min: Int = 0,
    max: Int = 10,
    step: Int = 1,
    label: String? = null,
    enabled: Boolean = true,
) {
    fun clamp(n: Int): Int = n.coerceIn(min, max)

    val atMin = value <= min
    val atMax = value >= max

    Column(modifier = modifier) {
        if (label != null) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelLarge,
            )
        }

        Row(verticalAlignment = Alignment.CenterVertically) {
            IconButton(
                onClick = {
                    val next = clamp(value - step)
                    if (next != value) onValueChange(next)
                },
                enabled = enabled && !atMin,
                colors = IconButtonDefaults.iconButtonColors(
                    contentColor = MaterialTheme.colorScheme.primary,
                ),
            ) {
                Icon(
                    imageVector = Icons.Default.Remove,
                    contentDescription = "Decrement",
                )
            }

            Text(
                text = value.toString(),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(horizontal = 16.dp),
                color = if (enabled) {
                    MaterialTheme.colorScheme.onSurface
                } else {
                    MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                },
            )

            IconButton(
                onClick = {
                    val next = clamp(value + step)
                    if (next != value) onValueChange(next)
                },
                enabled = enabled && !atMax,
                colors = IconButtonDefaults.iconButtonColors(
                    contentColor = MaterialTheme.colorScheme.primary,
                ),
            ) {
                Icon(
                    imageVector = Icons.Default.Add,
                    contentDescription = "Increment",
                )
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/form-controls/Stepper.swift
// ============================================================
// Clef Surface SwiftUI Widget — Stepper
//
// Compact increment/decrement control using SwiftUI Stepper
// or custom [-] value [+] layout. Respects min, max, and step.
// ============================================================

import SwiftUI

// --------------- Component ---------------

/// Stepper view for compact increment/decrement control.
///
/// - Parameters:
///   - value: Binding to the current integer value.
///   - min: Minimum allowed value.
///   - max: Maximum allowed value.
///   - step: Step increment value.
///   - label: Optional label text.
///   - enabled: Whether the stepper is enabled.
struct StepperView: View {
    @Binding var value: Int
    var min: Int = 0
    var max: Int = 10
    var step: Int = 1
    var label: String? = nil
    var enabled: Bool = true

    private var atMin: Bool { value <= min }
    private var atMax: Bool { value >= max }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let label = label {
                Text(label)
                    .font(.headline)
            }

            HStack(spacing: 0) {
                SwiftUI.Button(action: {
                    let next = Swift.max(value - step, min)
                    if next != value { value = next }
                }) {
                    Image(systemName: "minus")
                        .frame(width: 44, height: 44)
                        .foregroundColor(enabled && !atMin ? .accentColor : .gray)
                }
                .disabled(!enabled || atMin)
                .accessibilityLabel("Decrement")

                Text("\(value)")
                    .font(.headline)
                    .fontWeight(.bold)
                    .frame(minWidth: 48)
                    .multilineTextAlignment(.center)
                    .foregroundColor(enabled ? .primary : .gray)
                    .padding(.horizontal, 16)

                SwiftUI.Button(action: {
                    let next = Swift.min(value + step, max)
                    if next != value { value = next }
                }) {
                    Image(systemName: "plus")
                        .frame(width: 44, height: 44)
                        .foregroundColor(enabled && !atMax ? .accentColor : .gray)
                }
                .disabled(!enabled || atMax)
                .accessibilityLabel("Increment")
            }
        }
    }
}
